File: app/models/system_config.py
"""系统配置相关数据模型

定义系统配置项的数据模型。
"""

from sqlalchemy import Boolean, Column, String, Text, DateTime, Integer
from sqlalchemy.sql import func

from app.core.database import BaseModel


class SystemConfig(BaseModel):
    """系统配置模型"""
    __tablename__ = "system_configs"
    
    key = Column(String(100), unique=True, index=True, nullable=False, comment="配置键")
    value = Column(Text, nullable=False, comment="配置值")
    description = Column(Text, nullable=True, comment="配置描述")
    category = Column(String(50), nullable=False, default="general", comment="配置分类")
    
    # 配置属性
    is_public = Column(Boolean, default=False, nullable=False, comment="是否为公开配置")
    is_encrypted = Column(Boolean, default=False, nullable=False, comment="是否加密存储")
    is_active = Column(Boolean, default=True, nullable=False, comment="是否启用")
    
    # 数据类型
    data_type = Column(String(20), default="string", nullable=False, comment="数据类型")
    
    # 验证规则
    validation_rule = Column(Text, nullable=True, comment="验证规则（JSON格式）")
    default_value = Column(Text, nullable=True, comment="默认值")
    
    def __repr__(self):
        return f"<SystemConfig(key='{self.key}', category='{self.category}')>"
    
    def to_dict(self, include_sensitive=False):
        """转换为字典，可选择是否包含敏感信息"""
        data = {
            "id": self.id,
            "key": self.key,
            "description": self.description,
            "category": self.category,
            "is_public": self.is_public,
            "is_active": self.is_active,
            "data_type": self.data_type,
            "default_value": self.default_value,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
        }
        
        # 只有在需要敏感信息或配置为公开时才包含值
        if include_sensitive or self.is_public:
            data["value"] = self.value
        
        if include_sensitive:
            data["is_encrypted"] = self.is_encrypted
            data["validation_rule"] = self.validation_rule
        
        return data
    
    @classmethod
    def get_by_key(cls, db, key: str):
        """根据键获取配置"""
        return db.query(cls).filter(cls.key == key, cls.is_active == True).first()
    
    @classmethod
    def get_by_category(cls, db, category: str, include_inactive=False):
        """根据分类获取配置列表"""
        query = db.query(cls).filter(cls.category == category)
        if not include_inactive:
            query = query.filter(cls.is_active == True)
        return query.order_by(cls.key).all()
    
    @classmethod
    def get_public_configs(cls, db):
        """获取所有公开配置"""
        return db.query(cls).filter(
            cls.is_public == True,
            cls.is_active == True
        ).order_by(cls.category, cls.key).all()

File: frontend/src/services/systemConfig.ts
import { apiService, ApiResponse } from './api';

// 系统配置类型定义
export interface SystemConfig {
  id: number;
  key: string;
  value: string;
  description?: string;
  category: string;
  is_public: boolean;
  is_encrypted: boolean;
  is_active: boolean;
  data_type: string;
  validation_rule?: string;
  default_value?: string;
  created_at: string;
  updated_at: string;
}

// 公开配置类型（不包含敏感信息）
export interface PublicSystemConfig {
  id: number;
  key: string;
  value: string;
  description?: string;
  category: string;
  data_type: string;
  created_at: string;
  updated_at: string;
}

// 配置分类类型
export interface ConfigCategory {
  category: string;
  count: number;
  description?: string;
}

// 创建配置请求类型
export interface CreateConfigRequest {
  key: string;
  value: string;
  description?: string;
  category: string;
  is_public?: boolean;
  is_encrypted?: boolean;
  is_active?: boolean;
  data_type?: string;
  validation_rule?: string;
  default_value?: string;
}

// 更新配置请求类型
export interface UpdateConfigRequest {
  value?: string;
  description?: string;
  category?: string;
  is_public?: boolean;
  is_active?: boolean;
  data_type?: string;
  validation_rule?: string;
  default_value?: string;
  is_encrypted?: boolean;
}

// 批量更新请求类型
export interface BatchUpdateRequest {
  configs: Record<string, string>;
  category?: string;
}

// 导出请求类型
export interface ExportRequest {
  categories?: string[];
  include_encrypted?: boolean;
  format?: 'json' | 'yaml' | 'env';
}

// 导入请求类型
export interface ImportRequest {
  data: string;
  format?: 'json' | 'yaml' | 'env';
  overwrite?: boolean;
  category?: string;
}

class SystemConfigService {
  // 获取系统配置列表（需要管理员权限）
  async getConfigs(
    category?: string,
    isPublic?: boolean,
    includeInactive?: boolean
  ): Promise<ApiResponse<SystemConfig[]>> {
    const params: any = {};
    if (category) params.category = category;
    if (isPublic !== undefined) params.is_public = isPublic;
    if (includeInactive) params.include_inactive = includeInactive;

    return apiService.get('/system/config', params);
  }

  // 获取公开配置（不需要认证）
  async getPublicConfigs(
    category?: string
  ): Promise<ApiResponse<PublicSystemConfig[]>> {
    const params: any = {};
    if (category) params.category = category;

    return apiService.get('/system/config/public', params);
  }

  // 获取配置分类列表
  async getCategories(): Promise<ApiResponse<ConfigCategory[]>> {
    return apiService.get('/system/config/categories');
  }

  // 获取单个配置
  async getConfig(key: string): Promise<ApiResponse<SystemConfig>> {
    return apiService.get(`/system/config/${key}`);
  }

  // 创建配置
  async createConfig(data: CreateConfigRequest): Promise<ApiResponse<SystemConfig>> {
    return apiService.post('/system/config', data);
  }

  // 更新配置
  async updateConfig(
    key: string,
    data: UpdateConfigRequest
  ): Promise<ApiResponse<SystemConfig>> {
    return apiService.put(`/system/config/${key}`, data);
  }

  // 删除配置
  async deleteConfig(key: string): Promise<ApiResponse<any>> {
    return apiService.delete(`/system/config/${key}`);
  }

  // 批量更新配置
  async batchUpdateConfigs(
    data: BatchUpdateRequest
  ): Promise<ApiResponse<{ updated: string[]; errors: string[] }>> {
    return apiService.post('/system/config/batch-update', data);
  }

  // 导出配置
  async exportConfigs(
    data: ExportRequest
  ): Promise<ApiResponse<{ format: string; count: number; data: string }>> {
    return apiService.post('/system/config/export', data);
  }

  // 导入配置
  async importConfigs(
    data: ImportRequest
  ): Promise<ApiResponse<{ imported: string[]; errors: string[] }>> {
    return apiService.post('/system/config/import', data);
  }

  // 获取特定分类的配置（辅助方法）
  async getConfigsByCategory(category: string): Promise<SystemConfig[]> {
    try {
      const response = await this.getConfigs(category);
      return response.code === 200 ? response.data : [];
    } catch (error) {
      console.error(`Error fetching configs for category ${category}:`, error);
      return [];
    }
  }

  // 获取配置值（辅助方法）
  async getConfigValue(key: string, defaultValue?: string): Promise<string | undefined> {
    try {
      const response = await this.getConfig(key);
      if (response.code === 200) {
        return response.data.value;
      }
      return defaultValue;
    } catch (error) {
      console.error(`Error fetching config value for key ${key}:`, error);
      return defaultValue;
    }
  }

  // 设置配置值（辅助方法）
  async setConfigValue(
    key: string,
    value: string,
    description?: string
  ): Promise<boolean> {
    try {
      // 先尝试获取现有配置
      const existing = await this.getConfig(key);
      
      if (existing.code === 200) {
        // 更新现有配置
        const response = await this.updateConfig(key, { value, description });
        return response.code === 200;
      } else {
        // 创建新配置
        const response = await this.createConfig({
          key,
          value,
          description,
          category: 'general'
        });
        return response.code === 200;
      }
    } catch (error) {
      console.error(`Error setting config value for key ${key}:`, error);
      return false;
    }
  }

  // 获取AI配置（特定分类的辅助方法）
  async getAIConfigs(): Promise<SystemConfig[]> {
    return this.getConfigsByCategory('ai');
  }

  // 获取下载配置
  async getDownloadConfigs(): Promise<SystemConfig[]> {
    return this.getConfigsByCategory('download');
  }

  // 获取上传配置
  async getUploadConfigs(): Promise<SystemConfig[]> {
    return this.getConfigsByCategory('upload');
  }

  // 获取系统配置
  async getSystemConfigs(): Promise<SystemConfig[]> {
    return this.getConfigsByCategory('system');
  }

  // 获取用户界面配置
  async getUIConfigs(): Promise<SystemConfig[]> {
    return this.getConfigsByCategory('ui');
  }

  // 验证配置值格式（客户端验证）
  validateConfigValue(value: string, dataType: string, validationRule?: string): boolean {
    try {
      switch (dataType) {
        case 'integer':
          return !isNaN(parseInt(value)) && isFinite(parseInt(value));
        case 'float':
          return !isNaN(parseFloat(value)) && isFinite(parseFloat(value));
        case 'boolean':
          return ['true', 'false', '1', '0'].includes(value.toLowerCase());
        case 'json':
          JSON.parse(value);
          return true;
        case 'string':
        case 'text':
        default:
          // 如果有验证规则，使用正则表达式验证
          if (validationRule) {
            const regex = new RegExp(validationRule);
            return regex.test(value);
          }
          return true;
      }
    } catch (error) {
      return false;
    }
  }

  // 格式化配置值显示
  formatConfigValue(value: string, dataType: string): string {
    switch (dataType) {
      case 'boolean':
        return ['true', '1'].includes(value.toLowerCase()) ? '是' : '否';
      case 'json':
        try {
          return JSON.stringify(JSON.parse(value), null, 2);
        } catch {
          return value;
        }
      default:
        return value;
    }
  }
}

// 创建系统配置服务实例
export const systemConfigService = new SystemConfigService();
export default systemConfigService;